Run the first cell, reload(CTRL + R), then run the second cell

In [ ]:
!wget https://github.com/korakot/kora/releases/download/v0.10/py310.sh
!bash ./py310.sh -b -f -p /usr/local
!python -m ipykernel install --name "py310" --user

--2023-01-17 15:01:53--  https://github.com/korakot/kora/releases/download/v0.10/py310.sh
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/266951884/0d0623be-3dec-4820-9e7b-69a3a5a75ef7?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230117%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230117T150153Z&X-Amz-Expires=300&X-Amz-Signature=2f6a7c2a31dbe8848544c722aa7db189b8c93f93b0d51e913be6b1bbd0ad3432&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=266951884&response-content-disposition=attachment%3B%20filename%3Dpy310.sh&response-content-type=application%2Foctet-stream [following]
--2023-01-17 15:01:53--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/266951884/0d0623be-3dec-4820-9e7b-69a3a5a75ef7?X-Amz-Algorithm=AWS4-HMAC-S

In [2]:
import sys
print("version:", sys.version)

version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0]


# Задание 1

(**NB.** для запуска примеров кода нужен Python версии не ниже **3.10**, допускается использование других версий, в этом случае нужно самостоятельно избавиться от конструкции `match`).

Есть следующий код для [автоматического дифференцирования](https://en.wikipedia.org/wiki/Automatic_differentiation), в котором используются особенности системы типов языка `Python`: 

In [3]:
from dataclasses import dataclass
from typing import Union, Callable
from numbers import Number

@dataclass
class Dual:
    value: float
    d: float
    
    def __add__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value + o_value, self.d + o_d)
            case Number():
                return Dual(float(other) + self.value, self.d)

    def __mul__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value * o_value, self.value * o_d + self.d * o_value)
            case Number():
                return Dual(float(other) * self.value, float(other) * self.d)

    __rmul__ = __mul__  # https://docs.python.org/3/reference/datamodel.html#object.__mul__
    __radd__ = __add__  # https://docs.python.org/3/reference/datamodel.html#object.__radd__


def diff(func: Callable[[float], float]) -> Callable[[float], float]:
    return lambda x: func(Dual(x, 1.0)).d

Поддерживаются две операции - сложение и умножение. Применить можно так:

In [4]:
# Функция, которую будем дифференцировать
def f(x: float) -> float:
    return 5 * x * x + 2 * x + 2

f_diff = diff(f)

# значение производной в точке x = 2
f_diff(2)

22.0

## Задание 1.1 (5 баллов)

Какие недостатки вы видите в данной реализации? Реализуйте поддержку (полностью самостоятельно или модифицируя приведенный код):
- [унарных операций](https://docs.python.org/3/reference/datamodel.html#object.__neg__) 
- деления
- возведения в степень

Каким образом можно проверить корректность решения?  Реализуйте достаточный, по вашему мнению, набор тестов.

In [5]:
from dataclasses import dataclass
from typing import Union, Callable
from numbers import Number
import numpy as np

@dataclass
class Dual:
    value: float
    d: float
    
    def __add__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value + o_value, self.d + o_d)
            case Number():
                return Dual(float(other) + self.value, self.d)

    def __mul__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value * o_value, self.value * o_d + self.d * o_value)
            case Number():
                return Dual(float(other) * self.value, float(other) * self.d)

    __rmul__ = __mul__
    __radd__ = __add__

    def __abs__(self) -> "Dual":
        return Dual(abs(self.value), self.value / abs(self.value) * self.d)

    def __neg__(self) -> "Dual":
        return Dual(-self.value, -self.d)
    
    def __sub__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value - o_value, self.d - o_d)
            case Number():
                return Dual(self.value - float(other), self.d)
    
    def __rsub__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return self.__sub__(other, self)
            case Number():
                return Dual(float(other) - self.value, -self.d)

    def __pos__(self) -> "Dual":
        return Dual(self.value, self.d)

    def __pow__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value ** o_value, o_value * self.value ** (o_value - 1) * 
                        self.d + self.value ** o_value * o_d * np.log(self.value))
            case Number():
                return Dual(self.value ** float(other), float(other) * self.value ** (float(other) - 1) * self.d)
    
    def __rpow__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return self.__pow__(other, self)
            case Number():
                return Dual(float(other) ** self.value, float(other) ** self.value * np.log(float(other)) * self.d)
    
    def __truediv__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return Dual(self.value / o_value, (self.d * o_value - self.value * o_d) / o_value ** 2)
            case Number():
                return Dual(self.value / float(other), self.d / float(other))
    
    def __rtruediv__(self, other: Union["Dual", Number]) -> "Dual":
        match other:
            case Dual(o_value, o_d):
                return self.__truediv__(other, self)
            case Number():
                return Dual(float(other) / self.value, -float(other) * self.value ** -2 * self.d)

def diff(func: Callable[[float], float]) -> Callable[[float], float]:
    return lambda x: func(Dual(x, 1.0)).d

In [6]:
functions = [lambda x: -x * 42 + 2 * +x,
           lambda x: abs(x * 10),
           lambda x: (1 / (x * x)) + (x / x) - (x / 3),
           lambda x: x ** 3 - 2 ** x + 3 * x ** x,
           lambda x: abs(-x / 3) / (((7 ** x) ** x) ** (5 / +x))]

for f in functions:
    f_diff = diff(f)
    print(f_diff(2))

-40.0
10.0
-0.5833333333333333
29.545177444479563
-2.1782559186926773e-08


## Задание 1.2 (7 баллов)
Придумайте способ и реализуйте поддержку функций:
- `exp()`
- `cos()`
- `sin()`
- `log()`

Добавьте соответствующие тесты

In [7]:
def exp(arg: Union["Dual", Number]) -> Union["Dual", Number]:
    match arg:
        case Dual(value, d):
            return Dual(np.exp(arg.value), np.exp(arg.value) * arg.d)
        case Number():
            return np.exp(arg)

def cos(arg: Union["Dual", Number]) -> "Dual":
    match arg:
        case Dual(value, d):
            return Dual(np.cos(arg.value), -np.sin(arg.value) * arg.d)
        case Number():
            return np.cos(arg)

def sin(arg: Union["Dual", Number]) -> "Dual":
    match arg:
        case Dual(value, d):
            return Dual(np.sin(arg.value), np.cos(arg.value) * arg.d)
        case Number():
            return np.sin(arg)

def log(arg: Union["Dual", Number]) -> "Dual":
    match arg:
        case Dual(value, d):
            return Dual(np.log(arg.value), 1 / arg.value * arg.d)
        case Number():
            return np.log(arg)

In [8]:
functions_2 = [lambda x: sin(x ** 4),
           lambda x: exp(x / 4.1),
           lambda x: cos(x) ** 2,
           lambda x: -log(x * 5),
           lambda x: exp(log(sin(x)) / log(x))]

for f in functions_2:
    f_diff = diff(f)
    print(f_diff(2))

-30.64510337034831
0.3972529308831084
0.7568024953079283
-0.5
-0.4893585500123102


## Задание 1.3 (3 балла)

Воспользуйтесь методами **численного** дифференцирования для "проверки" работы кода на нескольких примерах. Например,  библиотеке `scipy` есть функция `derivative`. Или реализуйте какой-нибудь метод численного дифференцирования самостоятельно (**+5 баллов**)

In [9]:
pip install scipy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 34.2 MB/s eta 0:00:00


In [10]:
from scipy.misc import derivative

def f(x: float) -> float:
    return 5 * x * x + 2 * x + 2

derivative(f, 2.)

/tmp/ipykernel_1359/2371880734.py:6: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
  derivative(f, 2.)


22.0

In [13]:
funcs = [lambda x: -x * 42 + 2 * +x,
        lambda x: abs(x * 10),
        lambda x: (1 / (x * x)) + (x / x) - (x / 3),
        lambda x: x ** 3 - 2 ** x + 3 * x ** x,
        lambda x: abs(-x / 3) / (((7 ** x) ** x) ** (5 / +x)),
        lambda x: sin(x ** 4),
        lambda x: exp(x / 4.1),
        lambda x: cos(x) ** 2,
        lambda x: -log(x * 5),
        lambda x: exp(log(sin(x)) / log(x))]

for f in funcs:
    print(derivative(f, 2., dx=1e-6))

-40.00000000559112
10.00000000139778
-0.5833333333593593
29.54517744591101
-2.1782559189110845e-08
-30.6451033649513
0.39725293099479586
0.7568024953247265
-0.500000000069889
-0.4893585500154174


/tmp/ipykernel_1359/1154223897.py:13: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
  print(derivative(f, 2., dx=1e-6))


Моя реализация:

In [15]:
def num_diff(f: Callable[[float], float], x: float, dx: float) -> float:
    return (f(x + dx) - f(x)) / dx

for f in funcs:
    print(num_diff(f, 2., dx=1e-6))

-40.000000012696546
10.000000003174137
-0.5833331460092239
29.545202689718053
-2.1782458963771925e-08
-30.644978945171264
0.39725297940051973
0.7568031490379212
-0.4999998748367318
-0.48935884200407287


## Задание 1.4 (10 баллов)

Необходимо разработать систему автоматического тестирования алгоритма дифференцирования в следующем виде:
- реализовать механизм генерации "случайных функций" (например, что-то вроде такого: $f(x) = x + 5 * x - \cos(20 * \log(12 - 20 * x * x )) - 20 * x$ )
- сгенерировать достаточно большое число функций и сравнить результаты символьного и численного дифференцирования в случайных точках 

Генерацию случайных функций можно осуществить, например, двумя путями. 
1. Генерировать функцию в текстовом виде, зачем использовать встроенную функцию [eval](https://docs.python.org/3/library/functions.html#eval)

```python
func = eval("lambda x: 2 * x + 5")
assert func(42) == 89 
```

2. Использовать стандартный модуль [ast](https://docs.python.org/3/library/ast.html), который позволяет во время выполнения программы манипулировать [Абстрактным Синтаксическим Деревом](https://ru.wikipedia.org/wiki/%D0%90%D0%B1%D1%81%D1%82%D1%80%D0%B0%D0%BA%D1%82%D0%BD%D0%BE%D0%B5_%D1%81%D0%B8%D0%BD%D1%82%D0%B0%D0%BA%D1%81%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D0%B4%D0%B5%D1%80%D0%B5%D0%B2%D0%BE).
Например, выражение 

```python
func = lambda x: 2 * x + 5
```

Можно запрограммировать с помощью кода:

```python

expr = ast.Expression(
    body=ast.Lambda(
        args=ast.arguments(
            args=[
                ast.arg(arg='x')
            ],
            posonlyargs=[],
            kwonlyargs=[],
            kw_defaults=[],
            defaults=[]
        ),
        body=ast.BinOp(
            left=ast.BinOp(
                left=ast.Constant(value=2),
                op=ast.Mult(),
                right=ast.Name(id='x', ctx=ast.Load())
            ),
            op=ast.Add(),
            right=ast.Constant(value=5)
        )
    )
)

ast.fix_missing_locations(expr)

func = eval(compile(expr, filename="", mode="eval"))

assert func(42) == 89
```

При реализации нужно учитывать области допустимых значений функций.

In [25]:
from random import randint, choice

def generate_function(var_number: int = 4) -> Callable[[float], float]:
    func = "lambda x: "

    un_op = ['+', '-']
    bi_op = [" + ", " - ", " * ", " / ", " ** "]
    math_functions = ["abs(", "exp(", "cos(", "sin(", "log("]

    scripts = ["var", "number", "un_op", "func", "open_br"]
    close_br_count = 0
    var_count = 0

    while(var_count < var_number):
        match choice(scripts):
            case "var":
                func += 'x'
                scripts = ["bi_op"] + ["close_br"] * close_br_count
                var_count += 1
            case "number":
                func += str(randint(0, 10))
                scripts = ["bi_op"] + ["close_br"] * close_br_count
            case "un_op":
                func += choice(un_op)
                scripts = ["var"] * 3 + ["number", "func", "open_br"]
            case "bi_op":
                func += choice(bi_op)
                scripts = ["var"] * 3 + ["number", "func", "open_br"]
            case "func":
                func += choice(math_functions)
                close_br_count += 1
                scripts = ["var"] * 3 + ["number", "un_op", "func", "open_br"]
            case "open_br":
                func += '('
                close_br_count += 1
                scripts = ["var"] * 3 + ["number", "un_op", "func"]
            case "close_br":
                func += ')'
                close_br_count -= 1
                scripts = ["bi_op"]

    if close_br_count > 0:
        func += ')' * close_br_count

    return func


for i in range(10):
    str_f = generate_function()
    print(str_f)
    f = eval(str_f)

    for j in range(5):
        try:
            x = randint(-30, 30)

            f_diff = diff(f)
            my_result = f_diff(2)

            ver_result = derivative(f, 2., dx=1e-6)
            
            print(f"{x} {my_result:.7f} {ver_result:.7f}, delta= {ver_result-my_result}")
        except:
            continue

lambda x: 1 ** 6 * 3 - (exp(0 / x) * x * x) * exp(x)
24 -59.1124488 -59.1124488, delta= -4.1963232888519997e-10
26 -59.1124488 -59.1124488, delta= -4.1963232888519997e-10
14 -59.1124488 -59.1124488, delta= -4.1963232888519997e-10
-28 -59.1124488 -59.1124488, delta= -4.1963232888519997e-10
19 -59.1124488 -59.1124488, delta= -4.1963232888519997e-10
lambda x: x + x + (x) / x
21 2.0000000 2.0000000, delta= 2.795559339574538e-10
1 2.0000000 2.0000000, delta= 2.795559339574538e-10
-6 2.0000000 2.0000000, delta= 2.795559339574538e-10
6 2.0000000 2.0000000, delta= 2.795559339574538e-10
-17 2.0000000 2.0000000, delta= 2.795559339574538e-10
lambda x: exp(x) + x * x ** (+exp(x))
-21 3130.5746349 3130.5746350, delta= 1.1179281500517391e-07
30 3130.5746349 3130.5746350, delta= 1.1179281500517391e-07
-13 3130.5746349 3130.5746350, delta= 1.1179281500517391e-07
25 3130.5746349 3130.5746350, delta= 1.1179281500517391e-07
10 3130.5746349 3130.5746350, delta= 1.1179281500517391e-07
lambda x: 1 * 6 / x -

/tmp/ipykernel_1359/4105758416.py:60: DeprecationWarning: scipy.misc.derivative is deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. You may consider using findiff: https://github.com/maroba/findiff or numdifftools: https://github.com/pbrod/numdifftools
  ver_result = derivative(f, 2., dx=1e-6)
/tmp/ipykernel_1359/383660780.py:25: RuntimeWarning: invalid value encountered in log
  return Dual(np.log(arg.value), 1 / arg.value * arg.d)
/tmp/ipykernel_1359/383660780.py:27: RuntimeWarning: invalid value encountered in log
  return np.log(arg)


## Задание 1.5 (7 баллов)

Реализуйте поддержку функций нескольких аргументов. Например

```python
def f(x: float, y: float, z: float) -> float:
    return x * y + z - 5 * y  


f_diff = diff(f)

f_diff(10, 10, 10) # = [10, 5, 1]
```

In [26]:
def diff(func: Callable[[tuple[float, ...]], float]) -> Callable[[tuple[float, ...]], list[float]]:
    args_number = func.__code__.co_argcount

    def f_diff(*args: tuple[float, ...]) -> list[float]:
        lst_d = list()

        for i in range(args_number):
            new_args = list()
            for j in range(args_number):
                if j == i:
                    new_args.append(Dual(args[j], 1.0))
                else:
                    new_args.append(Dual(args[j], 0.0))
            
            lst_d.append(func(*new_args).d)

        return lst_d
    
    return f_diff

def f(x: float, y: float, z: float) -> float:
    return x * y + z - 5 * y

f_diff = diff(f)
f_diff(10, 10, 10)

def f(x: float, y: float) -> float:
    return cos(x) * y

f_diff = diff(f)
f_diff(10, 10)

[5.440211108893697, -0.8390715290764524]